File: WebViewApp-ObjC/ProtoCat3/Delegates/MobileAccessDelegate.m
//
//  MobileAccessDelegate.m
//  ProtoCat3
//
//  Copyright © 2021 Blue Cedar Networks. All rights reserved.
//  Licensed to Microsoft under Contract #7267038.
//

#import <Foundation/Foundation.h>
#import "MobileAccessDelegate.h"
#import "MobileAccessLogDelegate.h"
#import "IntuneDelegate.h"

const NSNotificationName kMobileAccessStatusUpdatedNotificationName = @"MobileAccessStatusUpdatedNotification";

@implementation MobileAccessDelegate

static MobileAccessDelegate *sm_sharedDelegate = NULL;

+ (instancetype)sharedDelegate
{
    if (NULL == sm_sharedDelegate)
    {
        sm_sharedDelegate = [MobileAccessDelegate new];
        sm_sharedDelegate.config = [NSMutableDictionary dictionary];
        [sm_sharedDelegate.config addEntriesFromDictionary: @{
            [NSString stringWithUTF8String:kLoggingClassMobileAccess]: [NSString stringWithUTF8String:kLoggingSeverityDebug],
            [NSString stringWithUTF8String:kLoggingClassConnect]: [NSString stringWithUTF8String:kLoggingSeverityDebug],
            [NSString stringWithUTF8String:kLoggingClassInternal]: [NSString stringWithUTF8String:kLoggingSeverityDebug]
        }];
    }
    return sm_sharedDelegate;
}

- (void)configureSDK
{
    self.m_api = MobileAccessAPI.sharedInstance;
    MobileAccessError err = [sm_sharedDelegate.m_api mobileAccessInitializeWithDelegate:sm_sharedDelegate logDelegate:MobileAccessLogDelegate.logDelegate config:self.config];
    if (NoError != err)
    {
        NSLog(@"Failed to initialize MobileAccessAPI!");
    }
}

- (void)setVpnConfiguration:(NSDictionary *)vpnConfig
{
    const char *kRandom = "Don't fail";
    for (NSString *key in [vpnConfig allKeys]) {
        if ([key isEqualToString:[NSString stringWithUTF8String:kConnectionType]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kConnectionName]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kServerAddress]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kPacUrl]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kProxyAddress]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kProxyPort]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kTrustedCertificates]] ||
            [key isEqualToString:[NSString stringWithUTF8String:kRandom]])
        {
            continue;
        }
        else
        {
            NSLog(@"Unexpected key in vpn config: %@", key);
            return;
        }
    }
    [self.config addEntriesFromDictionary:vpnConfig];
}

- (MobileAccessStatus)getStatus
{
    return [self.m_api getStatus];
}

- (NSString *)getStatusString
{
    return [self.m_api getStatusString];
}

- (NSString *)getVersionString
{
    return [self.m_api getVersionString];
}

- (void)connect
{
    [self.m_api connect];
}

- (void)disconnect
{
    [self.m_api disconnect];
}

- (void)onReceivedEvent:(MobileAccessStatus)event
{
    NSLog(@"%s event: %u", __PRETTY_FUNCTION__, event);

    [NSNotificationCenter.defaultCenter postNotificationName:kMobileAccessStatusUpdatedNotificationName object:nil];
}

- (void)onConnected
{
    NSLog(@"%s", __PRETTY_FUNCTION__);
}


- (void)onDisconnected
{
    NSLog(@"%s", __PRETTY_FUNCTION__);
}


- (void)onError:(MobileAccessError)error
{
    NSLog(@"%s: error: %u", __PRETTY_FUNCTION__, error);
}


- (void)onInitialized
{
    NSLog(@"%s", __PRETTY_FUNCTION__);

    [self connect];
}


- (void)onReconnecting
{
    NSLog(@"%s", __PRETTY_FUNCTION__);
}


- (void)onUserInteractionRequired
{
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)onTokenRequiredWithCallback:(TokenRequestCallback)tokenCallback withFailedToken:(NSString *)failedToken
{
    [IntuneDelegate.sharedDelegate onTokenRequiredWithCallback:tokenCallback withFailedToken:failedToken];
}



@end

File: WebViewApp-Swift/TunnelMAMTestApp2/MobileAccessDelegate.swift
//
// Copyright (c) Microsoft Corporation.  All rights reserved.
//

import Foundation
import MobileAccessApi
import SwiftUI

public protocol ConnectionListener {
    func onConnected() -> Void
    func onReconnecting() -> Void
    func onInitialized() -> Void
    func onDisconnected() -> Void
    func onError(_ error: MobileAccessError) -> Void
}

public class MobileAccessDelegate: NSObject, MobileAccessApi.MobileAccessDelegate {
    
    static var sharedDelegate: MobileAccessDelegate = {
        let delegate = MobileAccessDelegate()
        delegate.config.addEntries(from: [
            String.fromUtf8(kLoggingClassMobileAccess): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassConnect): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassInternal): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassHttp): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassPacket): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassSocket): String.fromUtf8(kLoggingSeverityDebug),
            String.fromUtf8(kLoggingClassIntune): String.fromUtf8(kLoggingSeverityDebug)
        ])
        return delegate
    }()
    
    var connectionListeners: Array<ConnectionListener> = []
    let config: NSMutableDictionary = NSMutableDictionary.init()
    var api: MobileAccessAPI?
    
    func setVpnConfiguration(vpnConfig: Dictionary<String, Any>) {
        self.config.addEntries(from: vpnConfig)
    }
    
    func configureSDK(){
        self.api = MobileAccessAPI.sharedInstance
        let error = MobileAccessDelegate.sharedDelegate.api?.mobileAccessInitialize(with: MobileAccessDelegate.sharedDelegate, logDelegate: LogDelegate(), config: (self.config as! [String : String]))
        if error != NoError {
            NSLog("Failed to initialize MobileAccessAPI!")
        }
    }
    
    public func registerConnectionListener(delegate: ConnectionListener) {
        connectionListeners.append(delegate)
    }
    
    public func onTokenRequired(callback tokenCallback: TokenRequestCallback!, withFailedToken failedToken: String!) {
        IntuneDelegate.sharedDelegate.onTokenRequiredWithCallback(tokenCallback: tokenCallback, failedToken: "")
    }
    
    public func onInitialized() {
        MobileAccessDelegate.sharedDelegate.connect()
        connectionListeners.forEach {
            $0.onInitialized()
        }
    }
    
    public func onConnected() {
        connectionListeners.forEach {
            $0.onConnected()
        }
    }
    
    public func onReconnecting() {
        connectionListeners.forEach {
            $0.onReconnecting()
        }
    }
    
    public func onUserInteractionRequired() {
        // implement later
    }
    
    public func onDisconnected() {
        connectionListeners.forEach {
            $0.onDisconnected()
        }
    }
    
    public func onError(_ error: MobileAccessError) {
        connectionListeners.forEach {
            $0.onError(error)
        }
    }

    func connect() {
        MobileAccessAPI.sharedInstance.connect()
    }
    
    func disconnect() {
        MobileAccessAPI.sharedInstance.disconnect()
    }
    
    func getStatus() -> MobileAccessStatus{
        return MobileAccessAPI.sharedInstance.getStatus()
    }
    
    func getStatusString() -> String {
        return MobileAccessAPI.sharedInstance.getStatusString()
    }
    
    public func onReceivedEvent(_ event: MobileAccessStatus) {
        // do nothing yet.
    }
}
